### DISCUSSION TITLE
/hyndsight/smape/

### DISCUSSION THREAD

**Discussion by robjhyndman** [robjhyndman/robjhyndman.com#182] [Announcements]
Errors on percentage errors 
 https://robjhyndman.com/hyndsight/smape/

**Comment by robjhyndman** [comment]
```header
Originally posted by Matt on 2014-04-17 20:53:42
````
I\'d like a better understanding of how the heavier penalty MAPE puts on
over forecasting is relevant for forecast evaluation and model
selection.

In some sense, I don\'t see the asymmetry- if we hold the actual value
fixed, MAPE for over forecasting and under forecasting of the same
absolute magnitude will be the same. E.g. for actual value 100,
forecasts of 50 and 150 give equivalent MAPE (50%). Doesn\'t this imply
that given an expected value for the actual observation of the forecast
horizon, MAPE treats over and under forecasting equally whenever the
magnitude of forecast error is the same?

We only get the asymmetry, it seems, if we hold the magnitude of
forecast error the same and vary the expected value for the actuals,
which doesn\'t seem practically relevant.

It\'s not true, in other words, that you can &quot;cheat&quot; by low-balling a
forecast in order to improve forecast MAPE; as long as that\'s the case,
what is the problem with using it, as it\'s not going to favor models
that under forecast over those that over forecast? (I\'m assuming here
that we don\'t need to worry about intermittent demand.)

Any direction here would be most appreciated; your blog has been an
invaluable resource in my business forecasting education.

**Reply by robjhyndman** [reply]
```header
Originally posted on 2014-04-18 04:15:55
````
I agree that it makes more sense to consider the case where the actual
stays the same and the forecasts vary, because we can\'t change actuals
we can only change forecasts.

**Reply by robjhyndman** [reply]
```header
Originally posted by Matt on 2014-04-18 16:53:54
````
Thanks, good to get some clarity here. It would be a shame to avoid a
simple metric like MAPE based on a misunderstanding. MASE is helpful
too, though in some cases one won\'t have a naive forecast to work with
(e.g. for the first period of a new product\'s sales).

**Comment by robjhyndman** [comment]
```header
Originally posted by Matt on 2014-04-17 21:04:40
````
I should add (and this is from your Armstrong reference) that it\'s true
that under forecasting has a maximum MAPE 100% (in the case where the
forecast is always zero), whereas over forecasting has no upper bound;
this is assuming that the forecast is always positive, of course. This
still seems to have limited significance to the question of whether one
should use MAPE in assessing forecasts, provided that zero forecasts are
not common in practice.

**Reply by robjhyndman** [reply]
```header
Originally posted on 2014-04-18 04:18:59
````
It\'s zero (or very small) actuals that is the issue, not zero
forecasts. They come up a lot. e.g., if you are trying to predict stock
returns.

**Reply by robjhyndman** [reply]
```header
Originally posted by Matt on 2014-04-18 16:54:58
````
Absolutely right, that was a slip on my part.

**Comment by robjhyndman** [comment]
```header
Originally posted by Chad Scherrer on 2014-04-24 02:24:16
````
For most applications of this, the values are positive, and it makes
sense to either use a model with a log link (as in a GLM) or to just
log-transform the response. So is there any reason to prefer MAPE over
some statistic (MSE or MAE, perhaps) of the residuals on the log scale?
If the big deal is having them as percentages, I guess you could do
something weird like use a base 1.01 for the log. Still seems more
sensible and less arbitrary than MAPE, which has no connection to the
loss function of any model I\'ve ever seen.

**Reply by robjhyndman** [reply]
```header
Originally posted on 2014-04-24 03:47:16
````
The advantage of MAPE is interpretability, especially in a business
context. MSE of logs is beyond most managers. The MAPE is connected to a
strange loss function. See Gneiting\'s JASA paper:
[http://dx.doi.org/10.1198/j\...](http://dx.doi.org/10.1198/jasa.2011.r10138)

**Comment by robjhyndman** [comment]
```header
Originally posted by edyhsgr on 2014-05-04 06:43:03
````
Why is MAPE typically used instead of Median Absolute Percent Error? Is
MAPE better?

**Comment by robjhyndman** [comment]
```header
Originally posted by Luis on 2015-01-14 12:12:23
````
Hi Rob, I would like to know if the function accuracy() works with
bats(). I\'m trying to use it but I got some errors.

**Reply by robjhyndman** [reply]
```header
Originally posted on 2015-01-15 03:29:15
````
It should work. If you think there is a problem, please submit a bug
report at
[https://github.com/robjhynd\...](https://github.com/robjhyndman/forecast/issues)
including a minimal reproducible example.

**Comment by robjhyndman** [comment]
```header
Originally posted by Simon on 2015-04-28 10:46:20
````
I am no Mathematician, but some time ago in wrestling with this problem
I have modified this statistic to nMAPE (for Normalised MPAE) where the
divisor becomes the maximum of Actual and Forecast.

**Reply by robjhyndman** [reply]
```header
Originally posted by Adam on 2015-09-21 12:22:29
````
I recently started thinking about doing this as well. From what I can
tell, this is also symmetric (using the example above abs(150-100)/150 =
0.33, abs(100-150)/150 = 0.33 and what I like about it is it is bounded
between (0,1) or (0,100) if you multiple by 100 (for positive
measurements such as my use case is). For me this an intuitive bound for
error. what has been your experience with this? Is there any literature
to support this?

**Comment by robjhyndman** [comment]
```header
Originally posted by cmos on 2015-09-14 19:57:43
````
In the original paper by Makridakis and also in the M-3 paper the
denominator of the sMAPE is multiplied by 2 whereas in your blog post
the numerator is multiplied by 2. Additionally, (Makridakis 1993)
nowhere mentions the term &quot;sMAPE&quot;. This term is only used in the M-3
paper.

**Reply by robjhyndman** [reply]
```header
Originally posted on 2015-09-14 22:47:55
````
1\. No it isn\'t. In the two papers you mention, the denominator is
DIVIDED by 2 which is equivalent to multiplying the numerator by 2.\
2. Yes, Makridakis didn\'t use the acronym &quot;sMAPE&quot; in 1993. That came
later.

**Reply by robjhyndman** [reply]
```header
Originally posted by cmos on 2015-09-14 23:02:53
````
You\'re right. I read that wrong.

**Comment by robjhyndman** [comment]
```header
Originally posted by Himani Wadhwa on 2015-10-23 18:36:00
````
Hi Rob!

What can be the expected value of MAPE for a dataset having nearly 50
observations? Like I am getting an error of around 33%. Is it fine?
Also,how should i proceed further in case I want to reduce the error?

**Reply by robjhyndman** [reply]
```header
Originally posted on 2015-10-26 00:41:15
````
The number of observations is not as important as how predictable is the
thing you are interested in. A good MAPE is one that is better than what
everyone else gets for the same forecast objective.

**Comment by robjhyndman** [comment]
```header
Originally posted by randomdude on 2015-11-10 08:05:58
````
Hi Rob,\
could you give me some advice on how to calculate the MASE for
time-series with multiple seasonalities.
[http://stats.stackexchange\....](http://stats.stackexchange.com/questions/180947/calculate-mase-for-time-series-with-multiple-seasonalities &quot;http://stats.stackexchange.com/questions/180947/calculate-mase-for-time-series-with-multiple-seasonalities&quot;)\
Thanks a lot for your input!

**Reply by robjhyndman** [reply]
```header
Originally posted on 2015-11-10 11:26:07
````
The only issue is how to choose the base forecast method used in the
scaling factor. I suggest you pick the shortest of the seasonal periods
and use it with a seasonal naive scaling factor.

**Comment by robjhyndman** [comment]
```header
Originally posted by quantweb on 2015-11-29 21:15:45
````
Rob.. was your position on metaselection (&quot;selection of model selection
methods&quot;) ? Especially if one can only calculate data dependent
mesuares like MAPE or MASE (not being able to calculate BIC or AIC
because the models are from different classes). Thanks!

**Reply by robjhyndman** [reply]
```header
Originally posted on 2015-11-29 22:09:59
````
When AIC is unavailable, I tend to use time series cross-validation:
[http://robjhyndman.com/hynd\...](http://robjhyndman.com/hyndsight/tscvexample/)

**Reply by robjhyndman** [reply]
```header
Originally posted by quantweb on 2015-11-29 22:21:18
````
Thanks Rob. When i said MAPE or MASE i meant as out of sample errors. So
i was thinking in using them as model selection strategy and make them
&quot;compete&quot;. I am trying to improve model selection before using any
out-of-sample forecast error bound.

**Comment by robjhyndman** [comment]
```header
Originally posted by Vikram Murthy on 2015-11-30 04:27:01
````
Mr Hyndman .. thanks for the post but the accuracy calculation (for
MAPE, MAE et al) ends with an &quot;Inf&quot; even if 1 of the values in the
data series is a 0 .. i do understand that according to the formula one
of the errors would result in a divide by 0 but by adding this Inf to
all the other errors, using MAPE becomes quite problematic. This is
handled well for MASE though but try explaining MASE to a management
thats been using MAPE for 10 years and swears by it :) ..can we have
some sort of mechanism to handle 0\'s in the time series ? appreciate
all the work you have put into the packages !

**Comment by robjhyndman** [comment]
```header
Originally posted by Branko Radovanovic on 2017-03-13 08:47:17
````
The trouble with MAPE is not only that it\'s asymmetric, but also that
it may distort the evaluation by making a worse forecast appear better.
An example:

A (Actual observations) = 100, 1000\
Fa (Forecast A) = 200, 1100\
Fb (Forecast B) = 100, 500

MAPE(A, Fa) = 55\
MAPE(A, Fb) = 25

By MAPE alone, it seems that forecast B is better than forecast A.
However, in most cases (load forecasting, sales forecasting), the
real-life cost of a forecast error is proportional to the absolute value
of the residual, and forecast A is actually much better.

If we adopt a different metric:

APE = 100 sum(\|yt - yt\^\|)/sum(yt)

\...it\'s an entirely different story:

APE(A, Fa) = 200 / 1100 = 18.1\
APE(A, Fb) = 500 / 1100 = 45.5

Now, I\'m calling it APE, but I don\'t really know how it\'s called -
it\'s like WAPE without the weights (i.e. with all weights set to 1), so
APE seems reasonable. It\'s simple and easy to intepret,
scale-invariant, works fine with zeroes, values close to zero, and
negative values too (if the denominator is changed to sum(\|yt\|)).

There\'s MASE, of course, but I\'d hate to explain how MASE works to a
client - I\'d much rather go with APE. (For that purpose, I\'d skip over
RMSE too - if one manages to reduce RMSE by, say, 10%, how much will it
save in the actual costs? Well, if the cost is more or less directly
proportional to absolute error - and it typically is - then it\'s
anyone\'s guess.)

A strange thing about MASE is that, if the observations are shuffled
(i.e. if they occur in a different order), and the forecasts are
rearranged accordingly, then MASE changes, possibly by a large margin.
That seems somewhat counter-intuitive: if we have two forecasts with the
same pairs of observations and predictions {yt, yt\^}, only in a
different order, then one would expect them to be evaluated as same
(although, to be fair, this is really a theoretical rather than
practical consideration).

I understand that when percentage (rather than absolute) errors are
important, MAPE may be better - but in other cases (that\'s nine times
out of ten, I\'d say), why e.g. APE isn\'t more widely used? (In
accuracy() too, perhaps?) Is there a reason to pick MASE over APE?

**Comment by robjhyndman** [comment]
```header
Originally posted by Fabrizio Sebastiani on 2018-09-03 19:47:30
````
I think that the alleged asymmetry of the MAPE pointed out by Armstrong
is a false problem. MAPE #is# symmetric, since it equally penalizes
underestimation and overestimation; that is, if \$y\_{t}=150\$, then a
prediction of \$\\hat{y}\_{t}=100\$ and a prediction of
\$\\hat{y}\_{t}=200\$ will be equally penalized by MAPE, and this is (I
think) good. The unintuitive aspect in Armstrong's thinking (and
throughout this blog post) is IMHO the idea that one should strive for a
measure in which \$y\_{t}\$ and \$\\hat{y}\_{t}\$ are interchangeable.
But I do not see why this should be desirable: in prediction, the
reference should always be the true value, and MAPE correctly considers
it as the reference (the gold standard) by putting it at the
denominator. True value and predicted value simply do not have the same
status. If I am allowed an analogy, true value is &quot;the law&quot; which
individual behaviours should comply with, and you want to evaluate
individual behaviours in terms of whether and how they comply with the
law: you don't evaluate how the law complies with individual behaviours,
you simply don't care.

A measure for evaluating the accuracy of prediction should be such that
that it can fairly compare two different predictors that attempt to
predict the same true value. Most if not all the measures that are
mentioned in this post as being proposed to &quot;amend&quot; MAPE look unsuitable
to me from this point of view; a suitable measure should equally
penalize \$\\hat{y}\_{t}=100\$ and \$\\hat{y}\_{t}=200\$ when
\$y\_{t}=150\$, and none of these measures (aside MAPE) seem to do this.
The only symmetry we should care about when measuring prediction
accuracy is the one between underestimation and overestimation, which is
(in the absence of prescriptions to the contrary) desirable.

I work in a field that is probably different from, but related to, from
the field in which the author of this post and those who have commented
on it work (my interest lies in measures for evaluating the accuracy of
prevalence estimators obtained via supervised machine learning \-- a
task we call &quot;quantification&quot;), and I landed on this page quite by
chance. If anyone is interested in my recent paper on evaluation
measures for quantification, just drop me a line at fabseb60@gmail.com .

**Comment by woodjmichael** [comment]
I never understood why $e = y - \hat y$. It seems more sensical that "overforecasting" ($\hat y > y$) should yield  $e>0$ and "underforecasting" ($\hat y < y$) should yield $e<0$.

**Comment by woodjmichael** [comment]
Sorry one more: how do you suggest choosing `m` when there are multiple seasonalities in a time series? In electric demand data we often see daily, weekly, and annually/bi-annually, with of course the first two being much better predictors of values with a day or week horizon.

**Reply by robjhyndman** [reply]
I normally use the smallest seasonal period for this purpose. So for hourly data, I would use 24
